PolyEngine/RenderingDevice/OpenGL/Src/Proxy/GLShaderProgram.cpp

```diff
@@ -97,24 +94,68 @@ void GLShaderProgram::Validate()
 	}
 }
 
-
 void GLShaderProgram::LoadShader(eShaderUnitType type, const String& shaderName)
+{
+	String rawCode = LoadTextFileRelative(eResourceSource::ENGINE, shaderName);
+
+	if (rawCode.IsEmpty())
+		return;
+		
+	Dynarray<String> includedFiles;
+	Dynarray<String> linesToProcess = rawCode.Split("\n");
+	StringBuilder sb;
+	size_t lineCounter = 0;
+	while(lineCounter < linesToProcess.GetSize())
+	{
+		String currLine = linesToProcess[lineCounter].GetTrimmed();
+
+		std::regex includeRegex(R"(^#pragma include\s*\"(\w.*)\")", std::regex::ECMAScript); // regex: ^#pragma include\s*"(\w.*)"
+		auto regex_begin = std::cregex_iterator(currLine.GetCStr(), currLine.GetCStr() + currLine.GetLength(), includeRegex);
+		auto regex_end = std::cregex_iterator();
+
+		if (regex_begin != regex_end)
+		{
+			for (std::cregex_iterator i = regex_begin; i != regex_end; ++i)
+			{
+				const std::cmatch& match = *i;
+
+				String includePath = SHADERS_INCLUDE_DIR + String(match[1].str().c_str());
+				ASSERTE(!includedFiles.Contains(includePath), "LoadShader failed, found recursive includes!");
+				includedFiles.PushBack(includePath);
+					
+				String rawInclude = LoadTextFileRelative(eResourceSource::ENGINE, includePath);
+				linesToProcess.Insert(lineCounter + 1, rawInclude.Split('\n'));
```

Comment:
This whole splitting doesn't seem too optimized. Maybe we can put an annotation in comment `// @fixme optimize this` To later revisit this during the refactor. Also the way we handle circular includes here is stragne, why fail if we can simply ommit the include as it's already included?. Just put a warning in console.